src/cmd/lib/gen-secrets.ts

```diff
@@ -0,0 +1,77 @@
+import $RefParser from '@apidevtools/json-schema-ref-parser'
+import { writeFileSync } from 'fs'
+import { dump } from 'js-yaml'
+import { omit } from 'lodash-es'
+import { gucci, loadYaml } from '../../common/utils'
+
+const jsonPathToObj = (jsonPath: string, obj, value: string) => {
+  const arr = jsonPath.split('.')
+  let tmpObj = obj
+  for (let i = 0; i < arr.length - 1; i++) {
+    const item = arr[i]
+    if (!tmpObj[item]) tmpObj[item] = {}
+    tmpObj = tmpObj[item]
+  }
+  tmpObj[arr[arr.length - 1]] = `{{ ${value} }}`
+}
+
+export const extractSecrets = (schema: any, parentAddress?: string): Array<string> => {
+  const schemaKeywords = ['properties', 'anyOf', 'allOf', 'oneOf']
+
+  return Object.keys(schema)
+    .flatMap((key) => {
+      const childObj = schema[key]
+      if (typeof childObj !== 'object') return false
+      if ('x-secret' in childObj) return parentAddress ? `${parentAddress}.${key}` : key
+      let address
+      if (parentAddress === undefined) {
+        address = schemaKeywords.includes(key) ? undefined : key
+      } else if (schemaKeywords.includes(key) || !Number.isNaN(Number(key))) address = parentAddress
+      else address = `${parentAddress}.${key}`
+      return extractSecrets(childObj, address)
+    })
+    .filter(Boolean)
+    .map((s: string) => s.replace(/^properties\./, ''))
+}
+
+export const extractSecretGenerators = (schema: any): Array<string> => {
+  return Object.keys(schema)
+    .flatMap((key) => {
+      const childObj = schema[key]
+      if (typeof childObj !== 'object') return false
+      if ('x-secret' in childObj) return childObj['x-secret'] !== '' ? childObj['x-secret'] : 'empty'
+      return extractSecretGenerators(childObj)
+    })
+    .filter(Boolean) as Array<string>
+}
+
+export const generateSecrets = async (): Promise<string> => {
+  const schema = loadYaml('values-schema.yaml')
+  const derefSchema = await $RefParser.dereference(schema)
+  const cleanSchema = omit(derefSchema, ['definitions', 'properties.teamConfig']) // FIXME: lets fix the team part later
+  const secretGenerators = extractSecretGenerators(cleanSchema)
+  const secretsValuesPath = extractSecrets(cleanSchema)
+  console.log(secretGenerators)
+  console.log(secretsValuesPath)
+
+  const obj = {}
+  for (let i = 0; i < secretGenerators.length; i++) {
+    if (secretGenerators[i] !== 'empty') jsonPathToObj(secretsValuesPath[i], obj, secretGenerators[i])
```

Comment:
so does this mean that if a generator is found it always sets a new password in the structure? I am now worried that  those will be merged later on, effectively setting a new password every time. Let's wait and check where these secrets are used...

Reply:
All passwords will be generated in bootstrap just once when there is no env folder. Then if it is in otomi helm chart, the passwords that are provided by user will replace the generated one in merge step. 
If user runs bootstrap again, no password will be generated anymore.